### ISSUE TITLE
Add Minibatch Discrimination Layer

### ISSUE THREAD

**Issue by mcgibbon** [fchollet/keras#3615]
I have an implementation of minibatch discrimination as described in [this paper](https://arxiv.org/abs/1606.03498) in the form of a Keras layer. It helps prevent GANs from collapsing to a single output, as outlined in the paper. Once I get around to adding documentation and tests, I can get a PR ready and outline the API, if this is something you're interested in adding. Let me know if this is too specialized of a layer.

**Comment by FedericoV** [comment]
I would be very interested in this.  I tried implementing a version myself and ran into some issues.

Could you share the preliminary code you have?  I'd be happy to beta test it.

**Comment by mcgibbon** [comment]
The preliminary code I have can be found [here](http://pastebin.com/EU8vgMcv). Thanks in advance for testing it out! It seems to be doing its job for what I'm working on. I have num_kernels set to 5 and kernel_dim set to 3 just before my discriminator output.

**Comment by fchollet** [comment]
Potentially interested in adding it, since GANs are gaining in popularity.
It would have to come with an example script, to demonstrate how to use it
and why use it! (a GAN, presumably).

**Comment by FedericoV** [comment]
I just copied the code and am running it now with some GANs on a fairly
large dataset - will report results later.  One thing: the openAI people
originally defined W in terms of some bandwidth parameters, which were
initialized looking at some statistics of the dataset.  In Keras, doing
data-dependent initialization is somewhat tricky without doing a few hacks.

**Comment by mcgibbon** [comment]
I'll start adding documentation, testing, and an example script, and submit a PR when it's ready (may take a little while).

**Comment by mcgibbon** [comment]
PR ready, #3677. Let me know if you want any revisions.

**Comment by FedericoV** [comment]
Hi,

a minor comment.  In the Theano version of the minibatch paper, the kernel
parameters are initialied with a test batch.  Have you tested to see if
this makes a big difference?  I am trying now to see if it's possible to do
it using K.function - but I don't know how to do automatically do it at
train time (I don't think it's even possible)

**Comment by mcgibbon** [comment]
In the tensorflow implementation [here](https://github.com/openai/improved-gan/blob/master/imagenet/discriminator.py#L88), the Dense transformation occurs on line 88, and they use [random initialization](https://github.com/openai/improved-gan/blob/master/imagenet/ops.py#L236) for the kernels.

In the Theano/Lasagne implementation ([1](https://github.com/openai/improved-gan/blob/master/mnist_svhn_cifar10/train_cifar_minibatch_discrimination.py#L66), [2](https://github.com/openai/improved-gan/blob/master/mnist_svhn_cifar10/nn.py#L132)), there's a more complicated initialization for W, but again it doesn't seem to be based off of the training data.

Are you sure you're not thinking of the virtual batch normalization, where "each example x is normalized based on the statistics collected on a reference batch of examples that are chosen once and fixed at the start of training, and on x itself"?

**Comment by FedericoV** [comment]
I was thinking of this:

https://github.com/openai/improved-gan/blob/master/mnist_svhn_cifar10/nn.py#L156

which is defined here:

https://github.com/openai/improved-gan/blob/master/mnist_svhn_cifar10/train_cifar_minibatch_discrimination.py#L75

and here:

https://github.com/openai/improved-gan/blob/master/mnist_svhn_cifar10/train_cifar_minibatch_discrimination.py#L100


and then executed here:

https://github.com/openai/improved-gan/blob/master/mnist_svhn_cifar10/train_cifar_minibatch_discrimination.py#L139

Data driven initialization is not really discussed in the paper though - so
I'm not sure if it's just a trick that's required to get the layers to
learn a good manifold quickly.

**Comment by pritamqu** [comment]
is there any MinibatchDiscrimination layer available in tensorflow or keras?
